[o1vm] Dynamic Selectors for MIPS

Fizzixnerd opened the pull request.

Fizzixnerd commented:
Should be ready for review;

I will add my own review covering stuff I'm not sure of.

dannywillems commented on o1vm/src/interpreters/mips/witness.rs:
```diff
@@ -82,6 +83,7 @@ pub struct Env<Fp, PreImageOracle: PreImageOracleT> {
     pub scratch_state: [Fp; SCRATCH_SIZE],
     pub halt: bool,
     pub syscall_env: SyscallEnv,
+    pub selectors: Vec<Fp>,
```
I suppose you can use a boolean instead of a field element.

dannywillems replied:
You can also keep simply the integer value of the current selector.

dannywillems commented on o1vm/src/interpreters/mips/interpreter.rs:
```diff
@@ -977,7 +982,27 @@ pub trait InterpreterEnv {
+    env.activate_selector(usize::from(instr) - N_MIPS_REL_COLS);
+
+    // Assert that exactly one selector is active at a time.
+    env.add_constraint(
```
I would not add this in the interpreter, but instead outside, when getting the constraints.

dannywillems commented on o1vm/src/interpreters/mips/interpreter.rs:
```diff
@@ -977,7 +982,27 @@ pub trait InterpreterEnv {
+        }) - Env::constant(1),
+    );
+
+    // Assert all selectors are either zero or one.
```
I would not add this in the interpreter, but instead outside, when getting the constraints.

dannywillems commented on o1vm/src/interpreters/mips/tests_helpers.rs:
```diff
@@ -89,6 +91,9 @@ where
         halt: false,
         // Keccak related
         syscall_env: SyscallEnv::default(),
+        selectors: Vec::from(std::array::from_fn::<Fp, N_MIPS_SEL_COLS, _>(|_| {
```
It might be ok for a first version, but I would simply keep track of the actual integer representing the instruction, and in the `main.rs` file, I would build the actual vectors of evaluation points.

dannywillems reviewed (commented):
Look at how it was done initially: https://github.com/o1-labs/proof-systems/blob/5b687813895fef22a5fc199103b70c3fb0d2360e/optimism/src/main.rs#L84.

I would go with this kind of version.

marcbeunardeau88 commented on o1vm/src/interpreters/mips/constraints.rs:
```diff
@@ -68,6 +68,10 @@ impl<Fp: Field> InterpreterEnv for Env<Fp> {
         }))
     }
 
+    // TODO: Make sure this is correct: only the witness needs to
+    // worry about this method.
+    fn activate_selector(&mut self, _selector_idx: usize) {}
```
It's not correct since it's empty
Or you mean you want to do this in a follow up ?

Fizzixnerd commented:
@dannywillems I deleted the code I added to the witness, as it's not clear it's needed now.  Will add back later if it is, but for now this just adds the selectors to the proof inputs.

Fizzixnerd commented on o1vm/src/interpreters/mips/constraints.rs:
```diff
@@ -72,6 +73,16 @@ impl<Fp: Field> InterpreterEnv for Env<Fp> {
     }
 
+    fn activate_selector(&mut self, selector_idx: usize) {
+        let mut acc : Self::Variable = -Expr::Atom(ExprInner::Constant(ConstantExpr::Atom(ConstantExprInner::Constant(kimchi::circuits::expr::ConstantTerm::<Fp>::Literal(Fp::one())))));
```
Is there a less verbose way of writing this?  it's just `-1`.

marcbeunardeau88 replied:
There is a From<u64> for expr

Fizzixnerd commented on o1vm/src/interpreters/mips/constraints.rs:
```diff
@@ -72,6 +73,16 @@ impl<Fp: Field> InterpreterEnv for Env<Fp> {
+        for i in 0..N_MIPS_SEL_COLS {
+            let var = self.variable(MIPSColumn::Selector(i));
+            acc += var.clone();
+            self.add_constraint(var.clone() * var.clone() - var);
```
Add that they are all either zero or one

Fizzixnerd commented on o1vm/src/interpreters/mips/constraints.rs:
```diff
@@ -72,6 +73,16 @@ impl<Fp: Field> InterpreterEnv for Env<Fp> {
+            acc += var.clone();
+            self.add_constraint(var.clone() * var.clone() - var);
+        }
+        self.add_constraint(acc);
```
Because `acc` starts at -1, this is zero if only one is one.

dannywillems commented on o1vm/src/interpreters/mips/witness.rs:
```diff
@@ -141,23 +143,8 @@ impl<Fp: Field, PreImageOracle: PreImageOracleT> InterpreterEnv for Env<Fp, PreI
-        }
-        */
+    fn variable(&self, column: Self::Position) -> Self::Variable {
+        u64::try_from(usize::from(column)).unwrap()
```
This is not correct:
`variable` must return the actual value saved at the position `column`.
If the column is a `SCRATCH_STATE`, we get the value from `self.scratch_state`. If the value is a `Selector(idx)`, you should return 1 if it is the index of the selector for the current instruction, otherwise 0.

dannywillems commented on o1vm/src/pickles/main.rs:
```diff
@@ -84,6 +85,14 @@ pub fn main() -> ExitCode {
+        curr_proof_inputs
+            .evaluations
+            .selector
+            .push(Fp::from(ark_ff::BigInt::from(
```
Note: you can simply use `Fp::from(instr as u64)`

dannywillems commented on o1vm/src/interpreters/mips/column.rs:
```diff
@@ -159,3 +171,9 @@ impl ColumnIndexer for Instruction {
+
+impl Instruction {
+    pub fn to_selector_column_idx(self) -> usize {
+        usize::from(self) - N_MIPS_REL_COLS
```
This is not correct. You get an underflow.

dannywillems replied:
This is correct. `usize::from(self)` is not the conversion from `EnumIter` into a `usize`, it is actually an implementation from Instruction to `usize` defined above. Sorry.

dannywillems commented on o1vm/src/interpreters/mips/constraints.rs:
```diff
@@ -72,6 +74,20 @@ impl<Fp: Field> InterpreterEnv for Env<Fp> {
     }
 
+    fn activate_selector(&mut self, _selector_idx: usize) {
+        let mut acc: Self::Variable = -Expr::Atom(ExprInner::Constant(ConstantExpr::Atom(
```
You can use `Self::Variable::one()` and import `ark_ff::One`.

dannywillems reviewed (changes requested):
Running with the nodes give errors:
```
[2024-10-02T17:29:43Z INFO  o1vm::interpreters::mips::witness] processing step=0 pc=00000ae88c insn=000802ba1f ips=0.00 pages=8598 mem=33.6 MiB name=_rt0_mips_linux
[2024-10-02T17:29:46Z INFO  o1vm::interpreters::mips::witness] Snapshot state in snapshot-state-0.json, step 0
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: TryFromIntError(())', o1vm/src/pickles/main.rs:93:81
```

dannywillems commented:
See https://github.com/o1-labs/proof-systems/pull/2649

dannywillems reviewed (approved):
As https://github.com/o1-labs/proof-systems/pull/2649/ has been approved, and has been built on top of this PR, I would approve this one, and merge right after 2649.

dannywillems commented:
Taking care of fixing conflicts
